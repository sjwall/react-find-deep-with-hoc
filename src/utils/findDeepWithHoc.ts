import {
  Children,
  isValidElement,
  type FC,
  type PropsWithChildren,
  type ReactNode,
} from 'react'
import {findDeep, type FindDeepResult} from './findDeep'
import {rebuildTree} from './rebuildTree'

export type FindDeepWithHocResult =
  | [FC<PropsWithChildren>, ReactNode]
  | [undefined, undefined]

/**
 * Finds a node deep in the tree and returns it with a HOC that wraps its parents.
 *
 * @param node The tree to search in.
 * @param predicate The callback to check if a node is a match.
 * @returns A tuple where the first item is the Wrapper component and the second is the found node.
 */
export function findDeepWithHoc(
  node: ReactNode,
  predicate: Parameters<typeof findDeep>[1],
): FindDeepWithHocResult {
  let result: FindDeepResult | undefined
  Children.toArray(node)
    .filter((child) => isValidElement(child))
    .find((child) => {
      result = findDeep(child, predicate)
      if (result) {
        return true
      }
      return false
    })
  if (result) {
    return [
      ({children}: PropsWithChildren) => rebuildTree(result!, children),
      result.found,
    ]
  }
  return [undefined, undefined]
}
