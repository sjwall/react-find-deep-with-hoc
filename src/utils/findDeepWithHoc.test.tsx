import {View, Text} from 'react-native'
import {render} from '@testing-library/react-native'
import {findDeepWithHoc} from './findDeepWithHoc'

describe('findDeepWithHoc', () => {
  test('finds deep', () => {
    const a = 'Hello'
    const b = <Text>{a}</Text>
    const ba = <Text>Other</Text>
    const c = (
      <View>
        {ba}
        {b}
      </View>
    )
    const tree = <View>{c}</View>

    const [Wrapper, found] = findDeepWithHoc(tree, (value) => value === 'Hello')

    expect(found).toEqual('Hello')
    // @ts-expect-error fail if undefined
    expect(render(<Wrapper>{found} World</Wrapper>).toJSON()).toMatchSnapshot()
  })

  test("doesn't find", () => {
    const a = 'Hello World'
    const b = <Text>{a}</Text>
    const ba = <Text>Other</Text>
    const c = (
      <View>
        {ba}
        {b}
      </View>
    )
    const tree = <View>{c}</View>

    const [Wrapper, found] = findDeepWithHoc(tree, (value) => value === 'Hello')

    expect(found).toBeUndefined()
    expect(Wrapper).toBeUndefined()
  })
})
