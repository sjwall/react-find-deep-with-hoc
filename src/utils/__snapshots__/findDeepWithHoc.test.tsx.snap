// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findDeepWithHoc finds deep 1`] = `
<View>
  <View>
    <Text>
      Hello
       World
    </Text>
  </View>
</View>
`;
